### ISSUE TITLE
StreamWriter keeps working after being disposed if `leaveOpen: true` is used

### ISSUE THREAD

**Issue by jozkee** [dotnet/runtime#89646]
In below snippet I was expecting an exception to be thrown but instead the StreamWriter keeps working:
```cs
var ms = new MemoryStream();
var sw = new StreamWriter(ms, leaveOpen: true);

sw.Write("Hello");
sw.Dispose();

sw.Write(" World");
sw.Flush();
Console.WriteLine(Encoding.UTF8.GetString(ms.ToArray())); // prints "Hello World".
```
If I change leaveOpen to false, it does throw as I expected.

_Originally posted by @Jozkee in https://github.com/dotnet/runtime/pull/89477#discussion_r1277944108_
            
cc @Neme12 @dersia

**Comment by dersia** [comment]
Yeah, this is why I did the change in the first place and then I found out, that there are test on Stream that fail, I think it was the test NullWriterWriteText, but it will be another 2 hours until I have access to my pc again, then I can tell for certain.

To me it does not make sense for an disposed Writer to be write able or readable from. this does not mean, that the underlying stream is disposed if it isn't closeable. For StreamWriter it is implemented in a way that keeps the Writer in an undisposed state, if the underlying stream is not closeable. and as mentioned before I wanted to fix it, but there are tests that test that and when I blamed that file I could see, that those test are some of the oldest (I think it was over 8 years ago) added when the test file was created in the first place, so I can imaging that they were copied when net core was branched or the feature was ported.
________________________________
From: David Cantú ***@***.***>
Sent: Friday, July 28, 2023 8:29:12 PM
To: dotnet/runtime ***@***.***>
Cc: dersia ***@***.***>; Mention ***@***.***>
Subject: [dotnet/runtime] StreamWriter keeps working after being disposed if `leaveOpen: true` is used (Issue #89646)


In below snippet I was expecting an exception to be thrown but instead the StreamWriter keeps working:

var ms = new MemoryStream();
var sw = new StreamWriter(ms, leaveOpen: true);

sw.Write("Hello");
sw.Dispose();

sw.Write(" World");
sw.Flush();
Console.WriteLine(Encoding.UTF8.GetString(ms.ToArray())); // prints "Hello World".

If I change leaveOpen to false, it does throw as I expected.

Originally posted by @Jozkee<https://github.com/Jozkee> in #89477 (comment)<https://github.com/dotnet/runtime/pull/89477#discussion_r1277944108>

cc @Neme12<https://github.com/Neme12> @dersia<https://github.com/dersia>

**Comment by Neme12** [comment]
Note: This works as expected on `StreamWriter`, i.e. it throws after being disposed, regardless of `leaveOpen`.

**Comment by dersia** [comment]
I don't think so. have you tried? or did you mean anything other than StreamWriter?
________________________________

**Comment by Neme12** [comment]
@dersia I meant `StreamReader`. I fixed the comment.

**Comment by Jozkee** [comment]
If we fix this someone somewhere will likely get broken, so we either don't fix it or we do and document it as a breaking change.

**Comment by dersia** [comment]
So you think we should make the breaking change as part of this ticket and net8? I would suggest do get this in and then create a new bug ticket and fix and break it for net9, this would give us time to see if there are many broken during net9 preview times. not sure if so close to rc, this is a good thing to do. I would also like to put a bit more thought into it also for the Reader (it isn't really throwing because there is no operation called on the reader that checks for dispose, for the writer this happens during flush).

I can also take care of that other ticket, since if have all the code for it and the PR for that other ticket could target milestone 9.
________________________________

**Comment by Neme12** [comment]
> I would also like to put a bit more thought into it also for the Reader (it isn't really throwing because there is no operation called on the reader that checks for dispose,

`StreamReader` does throw when disposed:
![image](https://github.com/dotnet/runtime/assets/11444821/fb52c673-e531-4803-8d7b-076df9048d47)

**Comment by Neme12** [comment]
> I would also like to put a bit more thought into it also for the Reader (it isn't really throwing because there is no operation called on the reader that checks for dispose, for the writer this happens during flush)

`StreamReader` does throw when disposed:
![image](https://github.com/dotnet/runtime/assets/11444821/7c9d73f8-10bd-461f-b2af-a4fda2a1f877)

**Comment by dersia** [comment]
hmmm interesting, I was pretty sure I wrote a test for this. maybe it was just for DisposeAsync. i will look at it, once I get home.
________________________________

**Comment by dersia** [comment]
> > I would also like to put a bit more thought into it also for the Reader (it isn't really throwing because there is no operation called on the reader that checks for dispose, for the writer this happens during flush)
> 
> `StreamReader` does throw when disposed: ![image](https://user-images.githubusercontent.com/11444821/256926081-7c9d73f8-10bd-461f-b2af-a4fda2a1f877.png)

So I looked into the code and found out, why the Reader tests aren't throwing, although your testing throws. And I am not sure, if what I found is a as bad sign as I think it is. So basically the test are not really testing StreamReader, they are testing an internal derived class called `NullStreamReader` https://github.com/dotnet/runtime/blob/4ed355098d50ea419931025ed466d0265c55b22f/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs#L1412 which overrides Dispose:
```cs
protected override void Dispose(bool disposing)
{
    // Do nothing - this is essentially unclosable.
}
```

So the Tests on `Stream.NullTest.cs` won't test any of the Dispose logic of `StreamReader`. So there are now 2 different things I am wondering about.

1. Should we remove the Dispose override on this class?
2. do we need to override DisposeAync() on this class, now that DisposeAsync was newly added to StreamReader?

Not sure which is the right option here. I think to be fully correct, one of the two need to happen as part of https://github.com/dotnet/runtime/pull/89477 and right now I am leaning more towards the latter.
Although, this makes all test that involve Reading after dispose useless.

**Comment by dersia** [comment]
> So you think we should make the breaking change as part of this ticket and net8? I would suggest do get this in and then create a new bug ticket and fix and break it for net9, this would give us time to see if there are many broken during net9 preview times. not sure if so close to rc, this is a good thing to do. I would also like to put a bit more thought into it also for the Reader (it isn't really throwing because there is no operation called on the reader that checks for dispose, for the writer this happens during flush). I can also take care of that other ticket, since if have all the code for it and the PR for that other ticket could target milestone 9.

As I was looking at this from my phone, I didn't realize that this is a separate issue, so don't mind this

**Comment by ViveliDuCh** [comment]
The consensus in #125189 is that the risk of breaking existing code outweighs the benefit given no active customer demand. Preserving the current behavior for backward compatibility.
